### DISCUSSION TITLE
Workflow: Unwrap sarcophagus

### DISCUSSION THREAD

**Discussion by adamgall** [sarcophagus-org/sarcophagus-v2-specs#15] [Ideas]
This post is to discuss and work through the details of the workflow involved when an embalmer fails to re-wrap their sarcophagus, and the dead man's switch is triggered, colloquially called "unwrapping a sarcophagus".

To set the scene:

* The system knows about all `n` archaeologists that belong to this sarcophagus.
  * Does this actually matter? Could the system operate effectively if it didn't have this "whitelist" of which archaeologists can participate in unwrapping?

**Comment by adamgall** [comment]
This comment is to describe the workflow

1. A point in time passes, at which a sarcophagus becomes eligible for unwrapping.
2. `m` of the total `n` archaeologists for a sarcophagus are active at that time, and are ready to participate in unwrapping.
3. Those `m` archaeologists call the `unwrap` function on the contract, passing in the sarcophagus identifier and their private keys to decrypt their encrypted shard. (The "encrypted shard" is the encrypted blob that holds a shard of a private key, that private key will ultimately be used to decrypt the outer layer of the payload on Arweave).
4. The smart contract verifies that the given private key is valid and correct.
    * System already has public key that was used to encrypt shard, so use same utility contract as in V1 to derive a public key from input private key and compare.
5. The smart contract verifies that the caller is within the set of `m` archaeologists (meaning, once `m` of the `n` have "unwrapped", no more are able to).
    * This can definitely use some more nuanced considerations. Do we want to allow more than (and reward with (the bounty / `n`)) `m` unwrappers, even though they're not necessary?
6. The smart contract distributes reward, emits event containing the private key.
7. If the caller is the `m`th archaeologist, perform necessary "completion" tasks on the sarcophagus, to mark it as complete.

**Reply by sethhrbek** [reply]
For #5 above -- I think we need to *try* to reward whatever archaeologists that do their job (call unwrap with correct priv key).  The issue with this model is we don't know how much to pay out until the resurrection window has expired, or until *all* archaeologists (i.e. 5/5) have called unwrap successfully.

**Reply by sethhrbek** [reply]
We could pay out `1/n` as the bounty. Then if enough archs participate in unwrapping for the bounty to be paid, but some dont (3/5), the remaining bounty could be able to be claimed post-resurrection window by archaeologists that did participate in unwrapping. There's some extra TX fees here which makes this approach not-so-great

**Reply by adamgall** [reply]
We're on the same page here.
